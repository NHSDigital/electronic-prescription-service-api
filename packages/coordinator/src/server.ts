import Hapi from "@hapi/hapi"
import routes from "./routes"
import HapiPino from "hapi-pino"
import {isLocal} from "./utils/environment"
import {
  reformatUserErrorsToFhir,
  rejectInvalidProdHeaders,
  switchContentTypeForSmokeTest
} from "./utils/server-extensions"

export const createServer = ({collectLogs}: {collectLogs?: boolean}): Hapi.Server => {
  const server = Hapi.server({
    port: 9000,
    host: "0.0.0.0",
    routes: {
      cors: true, // Won't run as Apigee hosted target without this
      payload: {
        parse: false
      }
    }
  })

  if (collectLogs) {
    routes.forEach((route) => {
      route.options = {
        log: {
          collect: true
        }
      }
    })
  }

  server.route(routes)

  server.ext([
    {type: "onRequest", method: rejectInvalidProdHeaders},
    {type: "onPreResponse", method: reformatUserErrorsToFhir},
    {type: "onPreResponse", method: switchContentTypeForSmokeTest}
  ])

  return server
}

const configureLogging = async (server: Hapi.Server) => {
  return await HapiPino.register(server, {
    // For non-local environments, dont pretty print to avoid spamming logs
    ...(isLocal() && {
      transport: {
        target: "pino-pretty",
        options: {
          colorize: true,
          minimumLevel: process.env.LOG_LEVEL || "info",
          levelFirst: true,
          messageFormat: true,
          timestampKey: "time",
          translateTime: true,
          singleLine: false,
          mkdir: true,
          append: true
        }
      }
    }),
    // Redact Authorization headers, see https://getpino.io/#/docs/redaction
    redact: ["req.headers.authorization"],
    wrapSerializers: false
  })
}

export const init = async (): Promise<void> => {
  const server = createServer({})
  await configureLogging(server)
  await server.start()
  server.log("info", `Server running on ${server.info.uri}`)
}
