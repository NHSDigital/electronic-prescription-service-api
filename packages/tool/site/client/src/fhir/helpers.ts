import * as fhir from "fhir/r4"
import {
  getEpsNumberOfRepeatsAllowedExtension,
  getLongFormIdExtension,
  getUkCoreNumberOfRepeatsIssuedExtension,
  RepeatInformationExtension,
  URL_EPS_NUMBER_OF_REPEATS_ALLOWED,
  URL_NUMBER_OF_REPEATS_ISSUED
} from "./customExtensions"
import {COURSE_OF_THERAPY_TYPE_CODES} from "./reference-data/valueSets"
import {getMedicationRequestResources} from "./bundleResourceFinder"
import {generateShortFormIdFromExisting} from "./generatePrescriptionIds"
import {convertMomentToISODate} from "../formatters/dates"
import moment from "moment"

export interface MedicationDispense extends fhir.MedicationDispense {
  contained: Array<MedicationRequest | fhir.PractitionerRole>
}

export interface MedicationRequest extends fhir.MedicationRequest{
  identifier: Array<fhir.Identifier>
  groupIdentifier: fhir.Identifier
}

export function getMedicationRequestLineItemId(medicationRequest: fhir.MedicationRequest): string {
  return medicationRequest.identifier[0].value
}

export function getMedicationDispenseLineItemId(medicationDispense: MedicationDispense): string {
  const containedMedicationRequest = medicationDispense.contained
    ?.find(resource => resource?.resourceType === "MedicationRequest") as MedicationRequest
  return containedMedicationRequest.identifier[0].value
}

export function getMedicationDispenseId(medicationDispense: fhir.MedicationDispense): string {
  return medicationDispense.identifier[0].value
}

export function getTotalQuantity(quantities: Array<fhir.Quantity>): fhir.Quantity {
  const units = quantities.map(quantity => quantity.unit)
  const values = quantities.map(quantity => quantity.value)
  if (new Set(units).size > 1) {
    throw new Error("Attempting to consolidate quantities with mismatched units")
  }
  return {
    ...quantities[0],
    value: values.reduce((a, b) => a + b)
  }
}

export function createIdentifier(): fhir.Identifier {
  return {
    system: "https://tools.ietf.org/html/rfc4122",
    value: crypto.randomUUID()
  }
}

export function requiresDispensingRepeatInformationExtension(medicationRequest: fhir.MedicationRequest): boolean {
  return medicationRequest.courseOfTherapyType.coding[0].code !== COURSE_OF_THERAPY_TYPE_CODES.ACUTE
}

// eslint-disable-next-line max-len
export function createDispensingRepeatInformationExtension(medicationRequest: fhir.MedicationRequest): RepeatInformationExtension {
  const [currentIssueNumber, endIssueNumber] = getCurrentIssueNumberAndEndIssueNumber(medicationRequest)
  return {
    url: "https://fhir.nhs.uk/StructureDefinition/Extension-EPS-RepeatInformation",
    extension: [
      {
        url: URL_NUMBER_OF_REPEATS_ISSUED,
        valueInteger: currentIssueNumber - 1
      },
      {
        url: URL_EPS_NUMBER_OF_REPEATS_ALLOWED,
        valueInteger: endIssueNumber - 1
      }
    ]
  }
}

export function getCurrentIssueNumberAndEndIssueNumber(medicationRequest: fhir.MedicationRequest): [number, number] {
  const ukCoreRepeatsIssuedExtension = getUkCoreNumberOfRepeatsIssuedExtension(medicationRequest.extension)

  const currentIssueNumber = (ukCoreRepeatsIssuedExtension ? ukCoreRepeatsIssuedExtension.valueUnsignedInt : 0) + 1
  const endIssueNumber = getEndIssueNumber(medicationRequest)

  return [currentIssueNumber, endIssueNumber]
}

function getEndIssueNumber(medicationRequest: fhir.MedicationRequest): number {
  if (medicationRequest.basedOn?.length) {
    return getEpsNumberOfRepeatsAllowedExtension(medicationRequest.basedOn[0].extension).valueInteger + 1
  } else if (medicationRequest.dispenseRequest?.numberOfRepeatsAllowed) {
    return medicationRequest.dispenseRequest?.numberOfRepeatsAllowed + 1
  } else {
    return 1
  }
}

export function updateBundleIds(bundle: fhir.Bundle): void {
  const firstGroupIdentifier = getMedicationRequestResources(bundle)[0].groupIdentifier
  const originalShortFormId = firstGroupIdentifier.value

  const newShortFormId = generateShortFormIdFromExisting(originalShortFormId)
  const newLongFormId = crypto.randomUUID()

  bundle.identifier.value = crypto.randomUUID()
  getMedicationRequestResources(bundle).forEach(medicationRequest => {
    medicationRequest.identifier[0].value = crypto.randomUUID()
    const groupIdentifier = medicationRequest.groupIdentifier
    groupIdentifier.value = newShortFormId
    getLongFormIdExtension(groupIdentifier.extension).valueIdentifier.value = newLongFormId
  })
}

export function updateValidityPeriod(bundle: fhir.Bundle): void {
  const start = convertMomentToISODate(moment.utc())
  getMedicationRequestResources(bundle).forEach(medicationRequest => {
    const validityPeriod = medicationRequest.dispenseRequest?.validityPeriod
    if (validityPeriod) {
      const oldStart = moment.utc(validityPeriod.start)
      const isExpired = oldStart < moment.utc()
      if (isExpired) {
        validityPeriod.start = start
        validityPeriod.end = convertMomentToISODate(moment.utc().add(1, "month"))
      }
    }
  })
}

export function isRepeatDispensing(bundle: fhir.Bundle): boolean {
  return getMedicationRequestResources(bundle)
    .flatMap(medicationRequest => medicationRequest.courseOfTherapyType.coding)
    .some(coding => coding.code === "continuous-repeat-dispensing")
}

export function orderBundleResources(r1: fhir.Resource, r2: fhir.Resource): number {
  return getSortIndex(r1) - getSortIndex(r2)
}

function getSortIndex(resource: fhir.Resource) {
  switch (resource.resourceType) {
    case "MessageHeader":
      return 0
    case "MedicationRequest":
      return 1
    case "Patient":
      return 2
    case "Practitioner":
      return 3
    case "PractitionerRole":
      return 4
    case "Organization":
      return 5
    case "HealthcareService":
      return 6
    case "Location":
      return 7
    case "Provenance":
      return 8
    default:
      return 0
  }
}
