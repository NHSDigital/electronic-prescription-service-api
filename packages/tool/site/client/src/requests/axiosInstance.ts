import axios, {InternalAxiosRequestConfig} from "axios"
import {v4 as uuidv4} from "uuid"

const axiosInstance = axios.create({
  validateStatus: () => true
})

axiosInstance.interceptors.request.use(
  (config: InternalAxiosRequestConfig) => {
    const requestId = uuidv4()
    config.headers["x-correlation-id"] = requestId

    return config
  },
  (error) => {
    return Promise.reject(Error(error))
  }
)

export {axiosInstance}
