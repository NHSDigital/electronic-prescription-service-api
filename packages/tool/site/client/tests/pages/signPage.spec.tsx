import {waitFor} from "@testing-library/react"
import {screen} from "@testing-library/dom"
import {v4} from "uuid"
import pretty from "pretty"
import * as React from "react"
import MockAdapter from "axios-mock-adapter"
import userEvent from "@testing-library/user-event"
import {readBundleFromFile} from "../messages"
import {AppContextValue} from "../../src"
import {renderWithContext} from "../renderWithContext"
import SignPage from "../../src/pages/signPage"
import {OperationOutcome} from "fhir/r4"
import {redirect} from "../../src/browser/navigation"
import {axiosInstance} from "../../src/requests/axiosInstance"
import {MomentInput} from "moment"
import {internalDev} from "../../src/services/environment"
import {MemoryRouter} from "react-router-dom"

jest.mock("uuid")
;(v4 as jest.Mock).mockImplementation(() => "test-uuid")

const baseUrl = "baseUrl/"
const context: AppContextValue = {baseUrl, environment: internalDev}

const prescriptionsUrl = `${baseUrl}prescriptions`
const editPrescriptionsUrl = `${baseUrl}prescribe/edit`
const signatureRequestUrl = `${baseUrl}sign/upload-signatures`

const prescriptionOrder = readBundleFromFile("prescriptionOrder.json")

jest.mock("moment", () => {
  const actualMoment = jest.requireActual("moment")
  return ({
    ...actualMoment,
    utc: (inp?: MomentInput, strict?: boolean) => actualMoment.utc(inp ?? "2021-11-13T10:57:13.000Z", strict)
  })
})

jest.mock("../../src/browser/navigation")

const mock = new MockAdapter(axiosInstance)

beforeEach(() => mock.reset())
afterEach(() => mock.reset())

test("Displays loading text while prescription data is being requested", async () => {
  mock.onGet(prescriptionsUrl).reply(function () {
    return new Promise(function (resolve) {
      setTimeout(function () {
        resolve([200, [prescriptionOrder]])
      }
      , 1000)
    })
  })
  const {container} = renderWithContext(<SignPage/>, context)
  await waitFor(() => screen.getByText("Retrieving prescription details."))

  expect(screen.getByText("Retrieving prescription details.")).toBeTruthy()
  expect(pretty(container.innerHTML)).toMatchSnapshot()
})

test("Displays prescription summary if prescription details are retrieved successfully", async () => {
  mock.onAny(prescriptionsUrl).reply(200, [prescriptionOrder])

  const container = await renderPage()

  expect(screen.getByText("Sign & Send")).toBeTruthy()
  expect(pretty(container.innerHTML)).toMatchSnapshot()
})

test("Displays loading text while prescription is being sent", async () => {
  mock.onAny(prescriptionsUrl).reply(200, [prescriptionOrder])

  const container = await renderPage()
  userEvent.click(screen.getByText("Sign & Send"))
  await waitFor(() => screen.getByText("Sending signature request."))

  expect(pretty(container.innerHTML)).toMatchSnapshot()
})

test("Redirects and displays link if signature request upload is successful", async () => {
  mock.onAny(signatureRequestUrl).reply(200, {
    redirectUri: "https://example.com/"
  })
  mock.onAny(prescriptionsUrl).reply(200, [prescriptionOrder])
  mock.onAny(editPrescriptionsUrl).reply(200, {
    redirectUri: ""
  })

  const container = await renderPage()
  userEvent.click(screen.getByText("Sign & Send"))
  await waitFor(() => screen.getByText("Upload Complete"))

  expect(redirect).toHaveBeenCalledWith("https://example.com/")

  const link = screen.getByRole<HTMLAnchorElement>("link")
  expect(link.text).toEqual("Proceed to the Signing Service")
  expect(link.href).toEqual("https://example.com/")
  expect(pretty(container.innerHTML)).toMatchSnapshot()
})

test("Displays error message if prepare errors present", async () => {
  const operationOutcome: OperationOutcome = {
    resourceType: "OperationOutcome",
    issue: [{
      severity: "fatal",
      code: "invalid",
      diagnostics: "Some error message"
    }]
  }
  mock.onAny(signatureRequestUrl).reply(200, {
    prepareErrors: [operationOutcome]
  })
  mock.onAny(prescriptionsUrl).reply(200, [prescriptionOrder])
  mock.onAny(editPrescriptionsUrl).reply(200, {
    redirectUri: ""
  })

  const container = await renderPage()
  userEvent.click(screen.getByText("Sign & Send"))
  await waitFor(() => screen.getByText("Error"))

  expect(pretty(container.innerHTML)).toMatchSnapshot()
})

test("Displays error message if redirect URI not present", async () => {
  const operationOutcome: OperationOutcome = {
    resourceType: "OperationOutcome",
    meta: {
      lastUpdated: "2022-10-21T13:47:00+00:00"
    },
    issue: [{
      severity: "fatal",
      code: "invalid",
      diagnostics: "Some error message"
    }]
  }
  mock.onAny(signatureRequestUrl).reply(400, operationOutcome)
  mock.onAny(prescriptionsUrl).reply(200, [prescriptionOrder])
  mock.onAny(editPrescriptionsUrl).reply(200, {
    redirectUri: ""
  })

  const container = await renderPage()
  userEvent.click(screen.getByText("Sign & Send"))
  await waitFor(() => screen.getByText("Error"))

  expect(pretty(container.innerHTML)).toMatchSnapshot()
})

async function renderPage() {
  const {container} = renderWithContext(<MemoryRouter><SignPage/></MemoryRouter>, context)
  await waitFor(() => screen.getByText("Prescription Summary"))
  return container
}
