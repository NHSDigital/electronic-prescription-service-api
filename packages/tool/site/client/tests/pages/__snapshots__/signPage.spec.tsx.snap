// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Displays error message if prepare errors present 1`] = `
"<h1 class="nhsuk-label-wrapper"><label class="nhsuk-label nhsuk-label--xl">Error</label></h1>
<div class="nhsuk-error-summary" tabindex="-1" role="alert" aria-labelledby="error-summary-title">
  <h2 class="nhsuk-error-summary__title" id="error-summary-title">Something went wrong while processing your request:</h2>
  <div class="nhsuk-error-summary__body">Unexpected response from server.</div>
</div>
<div class="">
  <fieldset class="nhsuk-fieldset">
    <legend class="nhsuk-fieldset__legend nhsuk-fieldset__legend--m">Request Details</legend>
    <dl class="nhsuk-summary-list">
      <div class="nhsuk-summary-list__row">
        <dt class="nhsuk-summary-list__key">Method</dt>
        <dd class="nhsuk-summary-list__value">POST</dd>
      </div>
      <div class="nhsuk-summary-list__row">
        <dt class="nhsuk-summary-list__key">URL</dt>
        <dd class="nhsuk-summary-list__value">baseUrl/sign/upload-signatures</dd>
      </div>
    </dl>
    <details class="nhsuk-details nhsuk-expander">
      <summary class="nhsuk-details__summary"><span class="nhsuk-details__summary-text">Request Headers</span></summary>
      <div class="nhsuk-details__text">
        <div class="sc-fmZSGO nBNSW"><button class="nhsuk-button sc-jMpmlX bERjXr" aria-disabled="false" type="submit">Copy</button><a class="nhsuk-button sc-jMpmlX bERjXr" role="button" aria-disabled="false" draggable="false" href="data:text/xml;charset=utf-8,%7B%0A%20%20%22Accept%22%3A%20%22application%2Fjson%2C%20text%2Fplain%2C%20*%2F*%22%2C%0A%20%20%22Content-Type%22%3A%20%22application%2Fx-www-form-urlencoded%22%2C%0A%20%20%22nhsd-identity-authentication-method%22%3A%20%22N3_SMARTCARD%22%2C%0A%20%20%22x-correlation-id%22%3A%20%22test-uuid%22%0A%7D" download="message">Download</a></div><pre class="sc-dYOLZc lnAxrk">{
  "Accept": "application/json, text/plain, */*",
  "Content-Type": "application/x-www-form-urlencoded",
  "nhsd-identity-authentication-method": "N3_SMARTCARD",
  "x-correlation-id": "test-uuid"
}</pre>
      </div>
    </details>
  </fieldset>
</div>
<div class="">
  <fieldset class="nhsuk-fieldset">
    <legend class="nhsuk-fieldset__legend nhsuk-fieldset__legend--m">Response Details</legend>
    <dl class="nhsuk-summary-list">
      <div class="nhsuk-summary-list__row">
        <dt class="nhsuk-summary-list__key">Status</dt>
        <dd class="nhsuk-summary-list__value">200 undefined</dd>
      </div>
    </dl>
    <details class="nhsuk-details nhsuk-expander">
      <summary class="nhsuk-details__summary"><span class="nhsuk-details__summary-text">Response Headers</span></summary>
      <div class="nhsuk-details__text">
        <div class="sc-fmZSGO nBNSW"><button class="nhsuk-button sc-jMpmlX bERjXr" aria-disabled="false" type="submit">Copy</button><a class="nhsuk-button sc-jMpmlX bERjXr" role="button" aria-disabled="false" draggable="false" href="data:text/xml;charset=utf-8,%7B%7D" download="message">Download</a></div><pre class="sc-dYOLZc lnAxrk">{}</pre>
      </div>
    </details>
    <details class="nhsuk-details nhsuk-expander">
      <summary class="nhsuk-details__summary"><span class="nhsuk-details__summary-text">Response Body</span></summary>
      <div class="nhsuk-details__text">
        <div class="sc-fmZSGO nBNSW"><button class="nhsuk-button sc-jMpmlX bERjXr" aria-disabled="false" type="submit">Copy</button><a class="nhsuk-button sc-jMpmlX bERjXr" role="button" aria-disabled="false" draggable="false" href="data:text/xml;charset=utf-8,%7B%0A%20%20%22prepareErrors%22%3A%20%5B%0A%20%20%20%20%7B%0A%20%20%20%20%20%20%22resourceType%22%3A%20%22OperationOutcome%22%2C%0A%20%20%20%20%20%20%22issue%22%3A%20%5B%0A%20%20%20%20%20%20%20%20%7B%0A%20%20%20%20%20%20%20%20%20%20%22severity%22%3A%20%22fatal%22%2C%0A%20%20%20%20%20%20%20%20%20%20%22code%22%3A%20%22invalid%22%2C%0A%20%20%20%20%20%20%20%20%20%20%22diagnostics%22%3A%20%22Some%20error%20message%22%0A%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%5D%0A%20%20%20%20%7D%0A%20%20%5D%0A%7D" download="message">Download</a></div><pre class="sc-dYOLZc lnAxrk">{
  "prepareErrors": [
    {
      "resourceType": "OperationOutcome",
      "issue": [
        {
          "severity": "fatal",
          "code": "invalid",
          "diagnostics": "Some error message"
        }
      ]
    }
  ]
}</pre>
      </div>
    </details>
  </fieldset>
</div>
<div class="sc-fmZSGO nBNSW"><button class="nhsuk-button nhsuk-button--secondary" aria-disabled="false" type="button">Back</button></div>"
`;

exports[`Displays error message if redirect URI not present 1`] = `
"<h1 class="nhsuk-label-wrapper"><label class="nhsuk-label nhsuk-label--xl">Error</label></h1>
<div class="nhsuk-error-summary" tabindex="-1" role="alert" aria-labelledby="error-summary-title">
  <h2 class="nhsuk-error-summary__title" id="error-summary-title">Something went wrong while processing your request:</h2>
  <div class="nhsuk-error-summary__body">Unexpected response from server.</div>
</div>
<div class="">
  <fieldset class="nhsuk-fieldset">
    <legend class="nhsuk-fieldset__legend nhsuk-fieldset__legend--m">Request Details</legend>
    <dl class="nhsuk-summary-list">
      <div class="nhsuk-summary-list__row">
        <dt class="nhsuk-summary-list__key">Method</dt>
        <dd class="nhsuk-summary-list__value">POST</dd>
      </div>
      <div class="nhsuk-summary-list__row">
        <dt class="nhsuk-summary-list__key">URL</dt>
        <dd class="nhsuk-summary-list__value">baseUrl/sign/upload-signatures</dd>
      </div>
    </dl>
    <details class="nhsuk-details nhsuk-expander">
      <summary class="nhsuk-details__summary"><span class="nhsuk-details__summary-text">Request Headers</span></summary>
      <div class="nhsuk-details__text">
        <div class="sc-fmZSGO nBNSW"><button class="nhsuk-button sc-jMpmlX bERjXr" aria-disabled="false" type="submit">Copy</button><a class="nhsuk-button sc-jMpmlX bERjXr" role="button" aria-disabled="false" draggable="false" href="data:text/xml;charset=utf-8,%7B%0A%20%20%22Accept%22%3A%20%22application%2Fjson%2C%20text%2Fplain%2C%20*%2F*%22%2C%0A%20%20%22Content-Type%22%3A%20%22application%2Fx-www-form-urlencoded%22%2C%0A%20%20%22nhsd-identity-authentication-method%22%3A%20%22N3_SMARTCARD%22%2C%0A%20%20%22x-correlation-id%22%3A%20%22test-uuid%22%0A%7D" download="message">Download</a></div><pre class="sc-dYOLZc lnAxrk">{
  "Accept": "application/json, text/plain, */*",
  "Content-Type": "application/x-www-form-urlencoded",
  "nhsd-identity-authentication-method": "N3_SMARTCARD",
  "x-correlation-id": "test-uuid"
}</pre>
      </div>
    </details>
  </fieldset>
</div>
<div class="">
  <fieldset class="nhsuk-fieldset">
    <legend class="nhsuk-fieldset__legend nhsuk-fieldset__legend--m">Response Details</legend>
    <dl class="nhsuk-summary-list">
      <div class="nhsuk-summary-list__row">
        <dt class="nhsuk-summary-list__key">Status</dt>
        <dd class="nhsuk-summary-list__value">400 undefined</dd>
      </div>
    </dl>
    <details class="nhsuk-details nhsuk-expander">
      <summary class="nhsuk-details__summary"><span class="nhsuk-details__summary-text">Response Headers</span></summary>
      <div class="nhsuk-details__text">
        <div class="sc-fmZSGO nBNSW"><button class="nhsuk-button sc-jMpmlX bERjXr" aria-disabled="false" type="submit">Copy</button><a class="nhsuk-button sc-jMpmlX bERjXr" role="button" aria-disabled="false" draggable="false" href="data:text/xml;charset=utf-8,%7B%7D" download="message">Download</a></div><pre class="sc-dYOLZc lnAxrk">{}</pre>
      </div>
    </details>
    <details class="nhsuk-details nhsuk-expander">
      <summary class="nhsuk-details__summary"><span class="nhsuk-details__summary-text">Response Body</span></summary>
      <div class="nhsuk-details__text">
        <div class="sc-fmZSGO nBNSW"><button class="nhsuk-button sc-jMpmlX bERjXr" aria-disabled="false" type="submit">Copy</button><a class="nhsuk-button sc-jMpmlX bERjXr" role="button" aria-disabled="false" draggable="false" href="data:text/xml;charset=utf-8,%7B%0A%20%20%22resourceType%22%3A%20%22OperationOutcome%22%2C%0A%20%20%22meta%22%3A%20%7B%0A%20%20%20%20%22lastUpdated%22%3A%20%222022-10-21T13%3A47%3A00%2B00%3A00%22%0A%20%20%7D%2C%0A%20%20%22issue%22%3A%20%5B%0A%20%20%20%20%7B%0A%20%20%20%20%20%20%22severity%22%3A%20%22fatal%22%2C%0A%20%20%20%20%20%20%22code%22%3A%20%22invalid%22%2C%0A%20%20%20%20%20%20%22diagnostics%22%3A%20%22Some%20error%20message%22%0A%20%20%20%20%7D%0A%20%20%5D%0A%7D" download="message">Download</a></div><pre class="sc-dYOLZc lnAxrk">{
  "resourceType": "OperationOutcome",
  "meta": {
    "lastUpdated": "2022-10-21T13:47:00+00:00"
  },
  "issue": [
    {
      "severity": "fatal",
      "code": "invalid",
      "diagnostics": "Some error message"
    }
  ]
}</pre>
      </div>
    </details>
  </fieldset>
</div>
<div class="sc-fmZSGO nBNSW"><button class="nhsuk-button nhsuk-button--secondary" aria-disabled="false" type="button">Back</button></div>"
`;

exports[`Displays loading text while prescription data is being requested 1`] = `
"<div style="text-align: center;">
  <h1 class="nhsuk-label-wrapper"><label class="nhsuk-label nhsuk-label--xl sc-hoLldG uAeGU">Loading</label></h1>
  <div class="nhsuk-loader"><span class="nhsuk-loader__spinner"></span><span class="nhsuk-loader__text">Retrieving prescription details.</span></div>
</div>"
`;

exports[`Displays loading text while prescription is being sent 1`] = `
"<div style="text-align: center;">
  <h1 class="nhsuk-label-wrapper"><label class="nhsuk-label nhsuk-label--xl sc-hoLldG uAeGU">Loading</label></h1>
  <div class="nhsuk-loader"><span class="nhsuk-loader__spinner"></span><span class="nhsuk-loader__text">Sending signature request.</span></div>
</div>"
`;

exports[`Displays prescription summary if prescription details are retrieved successfully 1`] = `
"<form>
  <h1 class="nhsuk-label-wrapper"><label class="nhsuk-label nhsuk-label--xl"><span>Prescription Summary</span>
      <figure class="nhsuk-image"><img class="nhsuk-image__img sc-gwsNht dNzavL" id="editPrescription" srcset="baseUrl/static/BlackTie_Bold_full_set_Pencil_SVG_Blue.svg" sizes="50px"></figure>
    </label></h1>
  <div class="sc-fmZSGO nBNSW"><button class="nhsuk-button" aria-disabled="false" type="button">Download this Prescription</button></div>
  <dl class="nhsuk-summary-list">
    <div class="nhsuk-summary-list__row">
      <dt class="nhsuk-summary-list__key">ID</dt>
      <dd class="nhsuk-summary-list__value">B407F7-A99968-4DB2AU</dd>
    </div>
    <div class="nhsuk-summary-list__row">
      <dt class="nhsuk-summary-list__key">Prescription Type Code</dt>
      <dd class="nhsuk-summary-list__value">1001</dd>
    </div>
    <div class="nhsuk-summary-list__row">
      <dt class="nhsuk-summary-list__key">Course Of Therapy</dt>
      <dd class="nhsuk-summary-list__value">Continuous long term (repeat dispensing)</dd>
    </div>
    <div class="nhsuk-summary-list__row">
      <dt class="nhsuk-summary-list__key">Issue Number</dt>
      <dd class="nhsuk-summary-list__value">1 of 6</dd>
    </div>
    <div class="nhsuk-summary-list__row">
      <dt class="nhsuk-summary-list__key">Authored On</dt>
      <dd class="nhsuk-summary-list__value">13-Nov-2021</dd>
    </div>
    <div class="nhsuk-summary-list__row">
      <dt class="nhsuk-summary-list__key">Effective Date</dt>
      <dd class="nhsuk-summary-list__value">03-Sep-2021</dd>
    </div>
    <div class="nhsuk-summary-list__row">
      <dt class="nhsuk-summary-list__key">Nominated Pharmacy ODS Code</dt>
      <dd class="nhsuk-summary-list__value">FER21</dd>
    </div>
    <div class="nhsuk-summary-list__row">
      <dt class="nhsuk-summary-list__key">Nominated Pharmacy Type</dt>
      <dd class="nhsuk-summary-list__value">Other (e.g. Community Pharmacy)</dd>
    </div>
  </dl><label class="nhsuk-label nhsuk-label--m">Patient</label>
  <dl class="nhsuk-summary-list">
    <div class="nhsuk-summary-list__row">
      <dt class="nhsuk-summary-list__key">Name</dt>
      <dd class="nhsuk-summary-list__value">CORY, ETTA (MISS)</dd>
    </div>
    <div class="nhsuk-summary-list__row">
      <dt class="nhsuk-summary-list__key">NHS Number</dt>
      <dd class="nhsuk-summary-list__value">944 930 5552</dd>
    </div>
    <div class="nhsuk-summary-list__row">
      <dt class="nhsuk-summary-list__key">Date of Birth</dt>
      <dd class="nhsuk-summary-list__value">04-Jan-1999</dd>
    </div>
    <div class="nhsuk-summary-list__row">
      <dt class="nhsuk-summary-list__key">Gender</dt>
      <dd class="nhsuk-summary-list__value">Female</dd>
    </div>
    <div class="nhsuk-summary-list__row">
      <dt class="nhsuk-summary-list__key">Address</dt>
      <dd class="nhsuk-summary-list__value">123 Dale Avenue<br>Long Eaton<br>Nottingham<br>NG10 1NP</dd>
    </div>
  </dl>
  <div class="nhsuk-table__panel-with-heading-tab">
    <h3 class="nhsuk-table__heading-tab">Medication</h3>
    <table class="nhsuk-table">
      <caption class="nhsuk-table__caption">Medication summary</caption>
      <thead class="nhsuk-table__head">
        <tr class="nhsuk-table__row">
          <th class="nhsuk-table__header" scope="col">Details</th>
          <th class="nhsuk-table__header" scope="col">Quantity</th>
          <th class="nhsuk-table__header" scope="col">Unit</th>
          <th class="nhsuk-table__header" scope="col">Dosage Instructions</th>
        </tr>
      </thead>
      <tbody class="nhsuk-table__body">
        <tr class="nhsuk-table__row">
          <td class="nhsuk-table__cell">
            <div><b>Methotrexate 10mg/0.2ml solution for injection pre-filled syringes</b></div>
          </td>
          <td class="nhsuk-table__cell">63</td>
          <td class="nhsuk-table__cell">tablet</td>
          <td class="nhsuk-table__cell">
            <div>10 milligram, Inject, Subcutaneous route, once weekly</div>
          </td>
        </tr>
      </tbody>
    </table>
  </div><label class="nhsuk-label nhsuk-label--m">Prescriber</label>
  <dl class="nhsuk-summary-list">
    <div class="nhsuk-summary-list__row">
      <dt class="nhsuk-summary-list__key">Name</dt>
      <dd class="nhsuk-summary-list__value">EDWARDS, Thomas (DR)</dd>
    </div>
    <div class="nhsuk-summary-list__row">
      <dt class="nhsuk-summary-list__key">Professional Codes</dt>
      <dd class="nhsuk-summary-list__value">SDS Role Profile ID - 100102238986<br>Spurious Code - 612345<br>SDS User ID - 3415870201<br>GMC Number - 4428981</dd>
    </div>
    <div class="nhsuk-summary-list__row">
      <dt class="nhsuk-summary-list__key">Telecom</dt>
      <dd class="nhsuk-summary-list__value">01234567890</dd>
    </div>
    <div class="nhsuk-summary-list__row">
      <dt class="nhsuk-summary-list__key">Organization</dt>
      <dd class="nhsuk-summary-list__value">SOMERSET BOWEL CANCER SCREENING CENTRE (A99968)</dd>
    </div>
    <div class="nhsuk-summary-list__row">
      <dt class="nhsuk-summary-list__key">Address</dt>
      <dd class="nhsuk-summary-list__value">MUSGROVE PARK HOSPITAL<br>TAUNTON<br>TA1 5DA</dd>
    </div>
    <div class="nhsuk-summary-list__row">
      <dt class="nhsuk-summary-list__key">Trust / CCG</dt>
      <dd class="nhsuk-summary-list__value">TAUNTON AND SOMERSET NHS FOUNDATION TRUST (RBA)</dd>
    </div>
  </dl>
  <div class="nhsuk-form-group"><label class="nhsuk-label nhsuk-label--s" id="signingOptions--label" for="signingOptions">Choose a sign in option</label>
    <div class="nhsuk-radios" id="signingOptions" value="N3_SMARTCARD" aria-label="Choose a sign in option">
      <div class="nhsuk-radios__item"><input class="nhsuk-radios__input" id="signingOptions-1" name="signingOptions" type="radio" value="N3_SMARTCARD" checked=""><label class="nhsuk-label nhsuk-radios__label" id="signingOptions-1--label" for="signingOptions-1">N3 Smartcard</label></div>
      <div class="nhsuk-radios__item"><input class="nhsuk-radios__input" id="signingOptions-2" name="signingOptions" type="radio" value="IOS"><label class="nhsuk-label nhsuk-radios__label" id="signingOptions-2--label" for="signingOptions-2">IOS</label></div>
      <div class="nhsuk-radios__item"><input class="nhsuk-radios__input" id="signingOptions-3" name="signingOptions" type="radio" value="FIDO2"><label class="nhsuk-label nhsuk-radios__label" id="signingOptions-3--label" for="signingOptions-3">FIDO2</label></div>
    </div>
  </div>
  <div class="sc-fmZSGO nBNSW"><button class="nhsuk-button" aria-disabled="false" type="submit">Sign &amp; Send</button><button class="nhsuk-button nhsuk-button--secondary" aria-disabled="false" type="button">Back</button></div>
</form>"
`;

exports[`Redirects and displays link if signature request upload is successful 1`] = `
"<h1 class="nhsuk-label-wrapper"><label class="nhsuk-label nhsuk-label--xl">Upload Complete</label></h1><label class="nhsuk-label">Use the link below if you are not redirected automatically.</label>
<div class="nhsuk-action-link"><a class="nhsuk-action-link__link" href="https://example.com/"><svg class="nhsuk-icon nhsuk-icon__arrow-right-circle" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" aria-hidden="true" height="36" width="36">
      <path d="M0 0h24v24H0z" fill="none"></path>
      <path d="M12 2a10 10 0 0 0-9.95 9h11.64L9.74 7.05a1 1 0 0 1 1.41-1.41l5.66 5.65a1 1 0 0 1 0 1.42l-5.66 5.65a1 1 0 0 1-1.41 0 1 1 0 0 1 0-1.41L13.69 13H2.05A10 10 0 1 0 12 2z"></path>
    </svg><span class="nhsuk-action-link__text">Proceed to the Signing Service</span></a></div>"
`;
